File: src/Server/UI/Context.cs
namespace Belin.Base.UI;

using System.Text.Json.Serialization;

/// <summary>
/// Defines contextual modifiers.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum Context {

	/// <summary>
	/// A danger.
	/// </summary>
	Danger,

	/// <summary>
	/// A warning.
	/// </summary>
	Warning,

	/// <summary>
	/// An information.
	/// </summary>
	Info,

	/// <summary>
	/// A success.
	/// </summary>
	Success
}

/// <summary>
/// Provides extension methods for contextual modifiers.
/// </summary>
public static class ContextExtensions {

	/// <summary>
	/// Gets the icon corresponding to the specified context.
	/// </summary>
	/// <param name="context">The context.</param>
	/// <returns>The icon corresponding to the specified context.</returns>
	public static string GetIcon(this Context context) => context switch {
		Context.Danger => "error",
		Context.Warning => "warning",
		Context.Info => "info",
		Context.Success => "check_circle"
	};

	/// <summary>
	/// Returns the Bootstrap representation of the specified context.
	/// </summary>
	/// <param name="context">The context.</param>
	/// <returns>The Bootstrap representation of the specified context.</returns>
	public static string ToBoostrap(this Context context) => context.ToString().ToLowerInvariant();
}

File: src/Client/UI/Context.ts
/**
 * Defines contextual modifiers.
 */
export const Context = Object.freeze({

	/**
	 * A danger.
	 */
	Danger: "Danger",

	/**
	 * A warning.
	 */
	Warning: "Warning",

	/**
	 * An information.
	 */
	Info: "Info",

	/**
	 * A success.
	 */
	Success: "Success"
});

/**
 * Defines contextual modifiers.
 */
export type Context = typeof Context[keyof typeof Context];

/**
 * Gets the icon corresponding to the specified context.
 * @param context The context.
 * @returns The icon corresponding to the specified context.
 */
export function getIcon(context: Context): string {
	switch (context) {
		case Context.Danger: return "error";
		case Context.Success: return "check_circle";
		case Context.Warning: return "warning";
		default: return "info";
	}
}

/**
 * Returns the CSS representation of the specified context.
 * @param context The context.
 * @returns The CSS representation of the specified context.
 */
export function toCss(context: Context): string {
	return context.toLowerCase();
}
